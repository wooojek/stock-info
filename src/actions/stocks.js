export const addStock = (stock) => ({
  type: 'ADD_STOCK',
  stock,
});

export const setStocks = (stocks) => ({
  type: 'SET_STOCKS',
  stocks,
});

export const startSetStocks = () => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const stocks = JSON.parse(localStorage.getItem('stocks')) || [];
      dispatch(setStocks(stocks));
      resolve();
    });
  };
}