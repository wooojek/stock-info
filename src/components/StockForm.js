import React from 'react';

export default class StockForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.stock ? props.stock.name : '',
      error: '',
    }
  }

  onNameChange = (e) => {
    const name = e.target.value.toUpperCase();
    this.setState(() => ({ name }))
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.name) {
      this.setState(() => ({ errorState: 'Please provide name.' }));
    } else {
      this.setState(() => ({ errorState: '' }));
      this.props.onSubmit({
        name: this.state.name,
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorState && (<p className="form__error">{this.state.errorState}</p>)}
        <input
          type="text"
          className="text-input"
          placeholder="Company symbol"
          autoFocus
          value={this.state.name}
          onChange={this.onNameChange}
        />
        <div>
          <button className="button">Track</button>
        </div>
      </form>
    )
  }
};