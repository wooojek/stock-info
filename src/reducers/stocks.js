const defaultStockState = [
  {
    name: 'Apple',
  },
  {
    name: 'Google',
  },
];

export default (state = defaultStockState, action) => {
  switch (action.type) {
    case 'ADD_STOCK':
      return [
        action.stock,
        ...state,
      ];
    default:
      return state;
  }
};