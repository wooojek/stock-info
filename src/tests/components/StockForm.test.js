import React from 'react';
import { shallow } from 'enzyme';
import StockForm from '../../components/StockForm';
import stocks from '../fixtures/stocks';

test('should render StockForm correctly', () => {
    const wrapper = shallow(<StockForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render StockForm correctly with data', () => {
    const wrapper = shallow(<StockForm stock={stocks[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
    const wrapper = shallow(<StockForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { },
    });
    expect(wrapper.state('errorState').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set name on input change', () => {
    let value = 'New Name';
    const wrapper = shallow(<StockForm />);

    value = value.toUpperCase();
    wrapper.find('input').at(0).simulate('change', {
        target: {
            value
        }
    });
    expect(wrapper.state('name')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<StockForm stock={stocks[0]} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { },
    });
    expect(wrapper.state('errorState')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        name: stocks[0].name,
    });
});