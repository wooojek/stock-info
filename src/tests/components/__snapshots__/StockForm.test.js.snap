// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render StockForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Company symbol"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Track
    </button>
  </div>
</form>
`;

exports[`should render StockForm correctly with data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Company symbol"
    type="text"
    value="Apple"
  />
  <div>
    <button
      className="button"
    >
      Track
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Company symbol"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Track
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide name.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Company symbol"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Track
    </button>
  </div>
</form>
`;
